tests/typecheck.egg

```diff
@@ -0,0 +1,97 @@
+; type checking for simply typed lambda calculus
+
+(datatype Type 
+  (TUnit) 
+  (TArr Type Type) ; t1 -> t2
+)
+
+(rule ( ; injectivity of ->
+  (= (TArr fr1 to1) (TArr fr2 to2))
+)(
+  (= fr1 fr2) (= to1 to2))
+)
+
+(datatype Expr 
+  (Lam String Type Expr) ; lam x : t . e
+  (App Expr Expr) 
+  (Var String) 
+  (Unit)
+)
+
+(datatype Ctx 
+  (Nil) 
+  (Cons String Type Ctx)
+)
+
+; ctx |- expr : type
+(function typeof (Ctx Expr) Type)
+
+; ctx |- () : unit
+(rewrite (typeof ctx (Unit)) (TUnit))
+
+; ctx; x: t |- x : t
+(rewrite (typeof (Cons x t ctx) (Var x)) t)
+
+; ctx |- e1 :- t1 -> t2
+; ctx |- e2 : t1
+; -----------------
+; ctx |- e1 e2 : t2
+
+(rule ( ; creating demand for the types of e1, e2
+  (= (typeof ctx (App e1 e2)) t)
+)( ; TODO ugly syntax
+  (= (typeof ctx e1) (typeof ctx e1))
+  (= (typeof ctx e2) (typeof ctx e2))
+))
+
+(rule ( ; only check ctx |- e1 e2 : t2 that has a demand
+    (= (typeof ctx (App e1 e2)) t)
+    (= (typeof ctx e1) (TArr t1 t2))
+    (= (typeof ctx e2) t1)
+)(
+    (= t t2)
+))
+
+; ctx |- x : t
+; ------------------ y != x 
+; ctx; y: t |- x : t
+(rule ( ; creating demand to check ctx |- x : t
+  (= (typeof (Cons y ty ctx) x) t)
+  ; (!= x y) TODO disequality not supported yet
```

Comment:
`x` is an expression (var), not a string. The same applies to the next rule.

The following rules will run after #15 is merged :tada: :

```
(rule ( ; creating demand to check ctx |- x : t
  (= (typeof (Cons y ty ctx) vx) t)
  (= vx (Var x))
  (!= x y)
)(
  (= (typeof ctx vx) (typeof ctx vx))
))

(rule ( ; only check ctx; y: t |- x : t that has a demand
    (= (typeof ctx xv) t1)
    (= (typeof (Cons y ty ctx) xv) t2)
    (= xv (Var x))
    (!= x y)
)(
    (= t1 t2)
))
```